source/blog/2024-11-05-hanami-220.html.markdown

```diff
@@ -0,0 +1,244 @@
+
+## Persistence pays off: Hanami’s new database layer
+
+Hanami’s new database layer gives you a clear home for all aspects of your database interactions, along with the means to build your own clean interface for your app’s business layer to consume.
+
+When you generate a new app, you’ll have a ready-to-go `DATABASE_URL` in your `.env`. Our default database for new apps is SQLite, with Postgres and MySQL also supported.
+
+You can create a migration with `bundle exec hanami generate migration`, and fill it in:
+
+```ruby
+ROM::SQL.migration do
+  change do
+    create_table :posts do
+      primary_key :id
+      column :title, :text, null: false
+    end
+  end
+end
+```
+
+Now you can migrate your database with `hanami db migrate`.
+
+After this, you can generate a new relation: `hanami generate relation posts`. Relations describe your low-level data sources. Here, this means your database tables. Relations are also your place to add reusable, chainable methods that you can use as the building blocks for expressive, higher-level queries. Add something simple to get started:
+
+```ruby
+module MyApp
+  module Relations
+    class Posts < MyApp::DB::Relation
+      schema :posts, infer: true
+
+      use :pagination
+      per_page 20
+
+      def order_by_latest
+        order(self[:id].desc)
+      end
+    end
+  end
+end
+```
+
+Relations are the primary component of your app’s database layer. While you can interact with them directly, it’s better to build a repo. With repos, you get to build your very own database interface, so you can better manage how your data is accessed across your app. You can build a post repo with `hanami generate repo posts`. Here you can define a method to return your latest posts, built using the methods in your relation:
+
+```ruby
+module MyApp
+  module Repos
+    class PostRepo < MyApp::DB::Repo
+      def latest(page:)
+        posts.order_by_latest.page(page).to_a
+      end
+    end
+  end
+end
+```
+
+You can include this repo as a dependency of any other class in your app, which is how you can access your data wherever you need. In a view, for example:
+
+```ruby
+module MyApp
+  module Views
+    module Posts
+      class Index < MyApp::View
+        include Deps["repos.post_repo"]
+
+        expose :posts do |page: 1|
+          post_repo.latest(page:)
+        end
+      end
+    end
+  end
+end
+```
+
+Repo methods return structs: plain old value objects, just data, with no live connection back to the database. This means you can be confident in passing them all around your app, knowing things like accidental n+1 queries are a thing of the past.
+
+You can customize these structs by creating your own matching classes. Make one for your posts with `hanami generate struct post`. Inside these classes, you can access any of the attributes selected in your repo’s corresponding database query.
```

Comment:
```suggestion
You can optionally customize these structs by creating your own matching classes. Make one for your posts with `hanami generate struct post`. Inside these classes, you can access any of the attributes selected in your repo’s corresponding database query.
```
Not sure if this is necessary but might be clarifying

Reply:
Yeah, I think it's good, added 👍🏼